experimental/formatted_string/src/formatted_string.rs

```diff
@@ -150,30 +182,38 @@ impl<F: Copy> FormattedStringBuilder<F> {
         string: &str,
         field: F,
     ) -> Result<&mut FormattedStringBuilder<F>, FormattedStringBuilderError> {
-        if !self.chars.is_char_boundary(pos) {
-            Err(FormattedStringBuilderError::PositionNotCharBoundary)
+        if pos > self.bytes.len() {
+            Err(FormattedStringBuilderError::IndexOutOfBounds(pos))
+        } else if !unsafe { from_utf8_unchecked(&self.bytes) }.is_char_boundary(pos) {
+            Err(FormattedStringBuilderError::PositionNotCharBoundary(
+                pos,
+                unsafe { String::from_utf8_unchecked(self.bytes.clone()) },
```

Comment:
nit: in the interest of reducing unsafes, both here and above we should do this as a two-step "construct `str` from `[u8]` with `str::from_utf8_unchecked()`" and when converting to an error just call `.to_owned()` or something.